### ISSUE TITLE
Running many blake2 hashes in a loop results in Ruby segfault due to GC

### ISSUE THREAD

**Issue by grempe** [franckverrot/blake2#3]
Hi, I am working on a pretty significant pull request to this gem that I would like to submit very soon.  However I have run into a showstopper issue with the gem as it is right now.  There is apparently some kind of a memory allocation or cleanup issue in the C extension code that causes a segfault if Ruby Garbage collection is turned on (as it generally always is). Here is some sample code I have created that demonstrates the issue every time on my machine (OS X El Capitan).  I also tried this under several versions of Ruby with the same outcome.  The segfault messages are always different since I think the segfault happens at random times based on when Ruby decides to do a GC sweep so they are pretty much useless as far as I can tell.  I suspect something is getting GC'ed while it is still in use.

```
#!/usr/bin/env ruby
require 'blake2'

# If GC is disabled this will run cleanly through
# thousands of iterations. If GC is running normally
# if will crash after a couple of thousand hashes pretty
# much every time. Uncomment this to allow clean runs.
#GC.disable

i = 0
100000.times do
  puts i if i % 1000 == 0
  Blake2.new(32, Blake2::Key.none).digest('abc', :to_hex)
  i += 1
end
```

Unless I uncomment the `GC.disable` line this fails after a few thousand hashes every time for me.

Unfortunately, I don't have much experience with C or Ruby C extensions.  Do you think you could take a look?

Cheers,

Glenn

**Comment by franckverrot** [comment]
Hey Glenn! I'll take a look ASAP (probably today but can't promise anything). I might know why this is happening. Thanks a lot for your future contribution then :-)

**Comment by grempe** [comment]
I've submitted the PR I mentioned.  Hopefully you can resolve the segfault issue with Garbage Collection.

**Comment by grempe** [comment]
Hi,

I just tested out the new release using the test code in your commit message.  I am still getting a segfault.

```
~/src$ gem list | grep blake
~/src$ gem install blake2
Fetching: blake2-0.2.0.gem (100%)
Building native extensions.  This could take a while...
Successfully installed blake2-0.2.0
1 gem installed
~/src$ irb
irb(main):001:0> require 'blake2'
true
irb(main):002:0> 1_000_000.times do
irb(main):003:1*       Blake2.new(32, Blake2::Key.none).digest('abc', :to_hex)
irb(main):004:1>     end
(irb):3: [BUG] rb_gc_mark(): 0x007f96210d40e8 is T_NONE
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]

-- Crash Report log information --------------------------------------------
   See Crash Report log file under the one of following:
     * ~/Library/Logs/CrashReporter
     * /Library/Logs/CrashReporter
     * ~/Library/Logs/DiagnosticReports
     * /Library/Logs/DiagnosticReports
   for more details.
Don't forget to include the above Crash Report log file in bug reports.

-- Control frame information -----------------------------------------------
c:0020 p:0030 s:0078 e:000076 BLOCK  (irb):3 [FINISH]
c:0019 p:---- s:0075 e:000074 CFUNC  :times
c:0018 p:0008 s:0072 e:000071 EVAL   (irb):2 [FINISH]
c:0017 p:---- s:0070 e:000069 CFUNC  :eval
c:0016 p:0025 s:0063 e:000062 METHOD /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/workspace.rb:87
c:0015 p:0027 s:0056 e:000054 METHOD /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/context.rb:380
c:0014 p:0024 s:0050 e:000049 BLOCK  /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:489
c:0013 p:0041 s:0042 e:000041 METHOD /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:623
c:0012 p:0011 s:0037 e:000036 BLOCK  /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:486
c:0011 p:0128 s:0033 e:000032 BLOCK  /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:246 [FINISH]
c:0010 p:---- s:0030 e:000029 CFUNC  :loop
c:0009 p:0009 s:0027 e:000026 BLOCK  /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:232 [FINISH]
c:0008 p:---- s:0025 e:000024 CFUNC  :catch
c:0007 p:0018 s:0021 e:000020 METHOD /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:231
c:0006 p:0037 s:0018 E:000dd0 METHOD /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:485
c:0005 p:0009 s:0015 e:000014 BLOCK  /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:395 [FINISH]
c:0004 p:---- s:0013 e:000012 CFUNC  :catch
c:0003 p:0174 s:0009 E:000ae0 METHOD /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:394
c:0002 p:0023 s:0004 E:0026e0 EVAL   /usr/local/var/rbenv/versions/2.3.1/bin/irb:11 [FINISH]
c:0001 p:0000 s:0002 E:000590 (none) [FINISH]

-- Ruby level backtrace information ----------------------------------------
/usr/local/var/rbenv/versions/2.3.1/bin/irb:11:in `<main>'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:394:in `start'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:394:in `catch'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:395:in `block in start'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:485:in `eval_input'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:231:in `each_top_level_statement'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:231:in `catch'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:232:in `block in each_top_level_statement'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:232:in `loop'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb:246:in `block (2 levels) in each_top_level_statement'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:486:in `block in eval_input'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:623:in `signal_status'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb:489:in `block (2 levels) in eval_input'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/context.rb:380:in `evaluate'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/workspace.rb:87:in `evaluate'
/usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/workspace.rb:87:in `eval'
(irb):2:in `irb_binding'
(irb):2:in `times'
(irb):3:in `block in irb_binding'

-- C level backtrace information -------------------------------------------
0   ruby                                0x0000000105784b14 rb_vm_bugreport + 388
1   ruby                                0x000000010561f0e5 rb_bug + 485
2   ruby                                0x0000000105642de6 gc_mark_children + 2358
3   ruby                                0x00000001056491db gc_mark_stacked_objects_incremental + 171
4   ruby                                0x0000000105646f76 newobj_slowpath + 1190
5   ruby                                0x0000000105646aa4 newobj_slowpath_wb_protected + 20
6   ruby                                0x0000000105703381 rb_str_resurrect + 65
7   ruby                                0x0000000105761122 vm_exec_core + 2802
8   ruby                                0x0000000105774d8c vm_exec + 124
9   ruby                                0x00000001057701a7 rb_yield_1 + 87
10  ruby                                0x000000010567a3f8 int_dotimes + 216
11  ruby                                0x000000010577a73a vm_call_cfunc + 314
12  ruby                                0x00000001057632a4 vm_exec_core + 11380
13  ruby                                0x0000000105774d8c vm_exec + 124
14  ruby                                0x0000000105782dee eval_string_with_cref + 1422
15  ruby                                0x0000000105770a61 rb_f_eval + 161
16  ruby                                0x000000010577a73a vm_call_cfunc + 314
17  ruby                                0x0000000105763500 vm_exec_core + 11984
18  ruby                                0x0000000105774d8c vm_exec + 124
19  ruby                                0x0000000105783260 loop_i + 48
20  ruby                                0x0000000105628cd8 rb_rescue2 + 280
21  ruby                                0x000000010577a73a vm_call_cfunc + 314
22  ruby                                0x00000001057632a4 vm_exec_core + 11380
23  ruby                                0x0000000105774d8c vm_exec + 124
24  ruby                                0x00000001057831f7 catch_i + 87
25  ruby                                0x0000000105771c0c rb_catch_protect + 156
26  ruby                                0x000000010577258c rb_f_catch + 76
27  ruby                                0x000000010577a73a vm_call_cfunc + 314
28  ruby                                0x00000001057632a4 vm_exec_core + 11380
29  ruby                                0x0000000105774d8c vm_exec + 124
30  ruby                                0x00000001057831f7 catch_i + 87
31  ruby                                0x0000000105771c0c rb_catch_protect + 156
32  ruby                                0x000000010577258c rb_f_catch + 76
33  ruby                                0x000000010577a73a vm_call_cfunc + 314
34  ruby                                0x00000001057632a4 vm_exec_core + 11380
35  ruby                                0x0000000105774d8c vm_exec + 124
36  ruby                                0x0000000105628078 ruby_exec_internal + 152
37  ruby                                0x0000000105627f86 ruby_run_node + 54
38  ruby                                0x00000001055da6af main + 79

-- Other runtime information -----------------------------------------------

* Loaded script: irb

* Loaded features:

    0 enumerator.so
    1 thread.rb
    2 rational.so
    3 complex.so
    4 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/x86_64-darwin15/enc/encdb.bundle
    5 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/x86_64-darwin15/enc/trans/transdb.bundle
    6 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/unicode_normalize.rb
    7 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/x86_64-darwin15/rbconfig.rb
    8 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/compatibility.rb
    9 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/defaults.rb
   10 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/deprecate.rb
   11 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/errors.rb
   12 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/version.rb
   13 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/requirement.rb
   14 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/platform.rb
   15 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/basic_specification.rb
   16 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/stub_specification.rb
   17 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/util/list.rb
   18 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/x86_64-darwin15/stringio.bundle
   19 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/specification.rb
   20 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/exceptions.rb
   21 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_gem.rb
   22 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/monitor.rb
   23 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb
   24 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems.rb
   25 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/path_support.rb
   26 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/rubygems/dependency.rb
   27 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/version.rb
   28 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/core_ext/name_error.rb
   29 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/levenshtein.rb
   30 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/jaro_winkler.rb
   31 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/spell_checkable.rb
   32 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/delegate.rb
   33 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/spell_checkers/name_error_checkers/class_name_checker.rb
   34 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/spell_checkers/name_error_checkers/variable_name_checker.rb
   35 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/spell_checkers/name_error_checkers.rb
   36 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/spell_checkers/method_name_checker.rb
   37 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/spell_checkers/null_checker.rb
   38 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean/formatter.rb
   39 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/did_you_mean-1.0.0/lib/did_you_mean.rb
   40 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/e2mmap.rb
   41 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/init.rb
#!/usr/bin/env ruby
   42 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/workspace.rb
   43 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/inspector.rb
   44 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/context.rb
   45 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/extend-command.rb
   46 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/output-method.rb
   47 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/notifier.rb
   48 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/slex.rb
   49 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-token.rb
   50 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/ruby-lex.rb
   51 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/src_encoding.rb
   52 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/magic-file.rb
   53 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/x86_64-darwin15/readline.bundle
   54 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/input-method.rb
   55 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/locale.rb
   56 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb.rb
   57 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/core_ext/array.rb
   58 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/core_ext/string.rb
   59 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/core_ext/method.rb
   60 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/core_ext/object.rb
   61 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/core_ext/class.rb
   62 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/core_ext/kernel.rb
   63 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/inspector.rb
   64 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/shellwords.rb
   65 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/formatter.rb
   66 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print/version.rb
   67 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/awesome_print-1.6.1/lib/awesome_print.rb
   68 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/2.3.0/irb/version.rb
   69 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/blake2-0.2.0/lib/blake2_ext.bundle
   70 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/blake2-0.2.0/lib/blake2/key.rb
   71 /usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/blake2-0.2.0/lib/blake2.rb

[NOTE]
You may have encountered a bug in the Ruby interpreter or extension libraries.
Bug reports are welcome.
For details: http://www.ruby-lang.org/bugreport.html

[1]    68901 abort      irb

/tmp$ ruby -v
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]
```

If I run it with the test script I provided in the bug report I get:

```
/tmp$ ./bl.rb
0
1000
2000
3000
4000
5000
ruby(69724,0x7fff789b2000) malloc: *** error for object 0x7fb1c2708750: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
[1]    69724 abort      ./bl.rb
```

And if I turn off GC in that script it will *usually* but not always make it through 100,000 runs.

So I think this issue is still open.

**Comment by grempe** [comment]
FYI, I am rebasing my commits against master and I see it when running rake test as well.

```
~/src/blake2 (fixes_20160513 ✗)$ rake test
/usr/local/var/rbenv/versions/2.3.1/bin/ruby -w -I"lib:test" -I"/usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rake-11.1.2/lib" "/usr/local/var/rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rake-11.1.2/lib/rake/rake_test_loader.rb" "test/**/*_test.rb"
Run options: --seed 39517

# Running:

.........ruby(81793,0x7fff789b2000) malloc: *** error for object 0x7ffd05313df0: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
rake aborted!
SignalException: SIGABRT

Tasks: TOP => test
(See full trace by running task with --trace)
```

**Comment by franckverrot** [comment]
Do you mind running `rake clean` too?

Can you please uninstall any version of `blake2` you might have (`gem uninstall` basically 😅), and use this command: `rake compile && ruby -Ilib -rblake2 test.rb` ?

Thanks for your help!

**Comment by grempe** [comment]
I'm out now. Will run again tonight.

**Comment by franckverrot** [comment]
Alright, I can reproduce it now.

**Comment by grempe** [comment]
Just for posterity, I did the steps above (and rake clobber too) and same result.  Glad you can reproduce now as well.

**Comment by franckverrot** [comment]
I've added a new test for the GC. `rake` will run it by default (100k iterations).  I won't say it it's fixed as you know... I failed to fix it yesterday, but please give it a shot.

I've replaced `ruby_xmalloc/free`s by regular `malloc/free`s.

**Comment by grempe** [comment]
Looking very good.  I modified my local test script to run 99_999_000 iterations.  I ran without fail.  I was running activity monitor watching that ruby process and it did grow to quite a size, almost 2GB before it exited cleanly.  So there is still perhaps some kind of slow memory growth.  No crashes though!! 👍 

I'd say ship it, and you might consider removing the last release from rubygems as it is known broken.

Thanks for doing the deep dive!
